models/block.go

```diff
@@ -73,11 +72,13 @@ func (b *Block) Hash() (gethCommon.Hash, error) {
 func decodeBlockEvent(event cadence.Event) (*Block, error) {
 	payload, err := events.DecodeBlockEventPayload(event)
 	if err != nil {
-		if block, err := decodeLegacyBlockEvent(event); err == nil {
-			return block, nil
-		}
+		return nil, err
+	}
 
-		return nil, fmt.Errorf("failed to cadence decode block [%s]: %w", event.String(), err)
+	var fixedHash *string
+	if payload.PrevRandao == gethCommon.HexToHash("0x0") {
+		hash := payload.Hash.String()
+		fixedHash = &hash
 	}
```

Comment:
@m-Peter @ramtinms I don't understand this part. Can you help me out?

Reply:
@janezpodhostnik This is a part which is necessary for calculating the hash of blocks, when there are changes to its format (e.g. addition/removal of fields). The hash calculation can only be made on demand, with a function (https://github.com/onflow/flow-go/blob/master/fvm/evm/types/block.go#L54-L63). When `testnet` was launched, the block format had certain fields. After some time, we made an HCU to introduce a new field, `PrevRandao`. This naturally caused the `Hash` function to return different values for past blocks. And this part is basically an approach to mitigate that. We fix the block hash, to the value that was emitted in the `EVM.BlockExecuted` event.